src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -463,15 +463,7 @@ internal void UpdateWithExemplar(long number, ReadOnlySpan<KeyValuePair<string,
                         this.runningValue.AsLong += number;
                     }
 
-                    if (isSampled)
-                    {
-                        Debug.Assert(this.mpComponents.ExemplarReservoir != null, "ExemplarReservoir was null");
-
-                        // TODO: Need to ensure that the lock is always released.
-                        // A custom implementation of `ExemplarReservoir.Offer` might throw an exception.
-                        this.mpComponents.ExemplarReservoir!.Offer(
-                            new ExemplarMeasurement<long>(number, tags));
-                    }
+                    this.OfferExemplarIfSampled(number, tags, isSampled);
```

Comment:
There should no point in calling `UpdateWithExemplar` method with `isSampled == false`.

@CodeBlanch Could you refactor the method signature first to remove `isSampled` as a parameter? We would also have to update the [call sites](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Metrics/AggregatorStore.cs#L996C1-L1000C14) to only call `UpdateWithExemplar` when the `ExemplarFilter` returns `true`.

Reply:
Not needed! The JIT is smart enough to do that when it inlines everything. Check it out: [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8lAbhvqfIDoARAS2wBzAHYRcGfmFztajZtwBKAVxGT8MbgGEI+AA78ANjCgBlEwDcpMGR0amAFtih6AMtmBLV6mLI4BmJlIGLQYAbxoGKKZA4hQGAFlyAAoAEwhlYGMGC2xDZRg0BkUYbFSAeRFDAE9TPWwRAB4AaRhqgDU8goAFbH4oRuY6IohgACsYMAwAfgA+WYYMYVwi4AgIQwZ+XFNsfWNUgEpwyOiz/gAzBmTt3f2YI9OzqIjqZ/fFh23ucouLk2SuXyhUWy0Osg+DAAvk9obDYcRYvEEqQ0hksjAcl0QSUypUanUGi02p1gb1+oNyMMGKMJlM5gslkIVgw1hstjs9noDsdXpCMF9cD8/iYAJIXO7ch6A7FFJks25cnkQ6JQs7wt7RADaACl+BgAOIwEQmKTJDDVPQwCAXNGZYyHQ4AXVhWoSMAFEFSov2yXdnu9+3KekkEBEQoAgkIhLBcLh+BYYKKqvwRKmhM7YXooAnsBhMXEGL9/lBxZKDo0ACqzZKVrHAoq4ipVWr1JqtDrY8kDIYjcaTGbzUHM1brTaK+6PTUvWHnK43TmTw6z6J8yHRS7XC1Wm2144AXn3i0t1tthjDGeX0/Xa/XzwF32LAOS55EGeStIHhyBBTlYJVd4wtekIwIYuCYpu5onrulYHke26nnaGKOiu7y3ne0QPkKT5QMkSEOh+/ZTN+srDrg4Koc8QEYQwoHgZRZzoTRvAwMAyhCNwABifSGMkABEACqJoIFaUwPMeVp8RRwEYVhwolskOgiImUAYNwlYQLw6LGDKwKHH+zLSTR1EfCZURmQwrr+g4Xo+tyfoejZgbcsGobhtwAByEDJoYqbppm17Zrm+ZMPEOH4ZiP44qUzYEm2xKdmSfQ9lSfZ0oOjJgrC6FAVCQA)

Reply:
Why have the extra code though? We never want to call `ExemplarReservoir.Offer` when `isSampled == false`.

Reply:
Oh sorry I thought you were talking about the call sites in `UpdateWithExemplar`. I agree with you and I plan to completely remove `UpdateWithExemplar` by the time I'm done with this effort 😄 That's what is on #5364. But let me get this in and then I'll keep getting there incrementally.

Reply:
Actually had to revert it. Things broke: https://github.com/open-telemetry/opentelemetry-dotnet/actions/runs/8088019825/job/22101191749?pr=5399

Doing this...

```diff
            if (this.IsExemplarEnabled() && this.exemplarFilter.ShouldSample(value, tags))
            {
                this.metricPoints[index].UpdateWithExemplar(value, tags: default);
            }
            else
            {
                this.metricPoints[index].Update(value);
            }
```

...seems to cause the unsampled stuff to flow with different locking mechanics. Yikes!

I do want to tackle this, but not here. So I'll give you two options:

1) We can merge this incremental refactor and I can get back to it.
2) We can close this. May or may not lead to a bigger PR in the future. The point of these small refactors was to reduce code duplication in order to make future refactoring PRs smaller 😄

Reply:
> ...seems to cause the unsampled stuff to flow with different locking mechanics. Yikes!

That's exactly what we want. Why should measurements that are unsampled by the `ExemplarFilter` take the slower update path (which takes locks) instead of the faster update path (which uses Interlocked statements). The slower path made sense only when we had to call `ExemplarReservoir.Offer` in the update logic.

I tried this refactoring in #5401 and it seems to work.

Reply:
> That's exactly what we want.

I take it back. Actually, it's not what we want 😄. That would cause issues when there are two concurrent updates: one which gets sampled and one that does not.

If at all the user has enabled exemplars, we have to take the slow path.

Reply:
Ah there was an issue with my refactor. I see it now why it broke things. But I still think it is a bad idea.

Consider the case of like `DoubleSumIncomingDelta` we'll have some threads running this (when exemplar is not sampled)...

```csharp
InterlockedHelper.Add(ref this.runningValue.AsDouble, number);
```

...and some threads running this (when exemplar is sampled)...

```csharp
                    this.mpComponents!.AcquireLock();

                    unchecked
                    {
                        this.runningValue.AsDouble += number;
                    }

                    this.OfferExemplarIfSampled(number, tags, isSampled);

                    this.mpComponents.ReleaseLock();
```

That second block is really this...

```csharp
                    this.mpComponents!.AcquireLock();

                    unchecked
                    {
                        var temp = this.runningValue.AsDouble;
                        this.runningValue.AsDouble = temp + 1;
                    }

                    this.OfferExemplarIfSampled(number, tags, isSampled);

                    this.mpComponents.ReleaseLock();
```

Imagine this case...

```csharp
                    this.mpComponents!.AcquireLock();

                    unchecked
                    {
                        var temp = this.runningValue.AsDouble;
                        // Thread here yields for a while. A bunch of InterlockedHelper.Add(ref this.runningValue.AsDouble, number) calls fire in the meantime.
                        this.runningValue.AsDouble = temp + 1; // We have now undone the interlocked calls
                    }

                    this.OfferExemplarIfSampled(number, tags, isSampled);

                    this.mpComponents.ReleaseLock();
```

See the comment about the yield. If you have some threads updating the value not using the lock, and some threads expecting the lock, we're going to have a race. No?

Reply:
> If at all the user has enabled exemplars, we have to take the slow path. We need to keep isSampled parameter for now. Sorry for the confusion.

No worries! I wish we could do it 😄 My goal is to get us there somehow!